src/DataStructures/StaticDeque.hpp

```diff
@@ -0,0 +1,520 @@
+    const auto old_end = this->begin() + old_size;
+    std::rotate(new_position, old_end, this->end());
+    return new_position;
+  } else {
+    // We are about to invalidate position.
+    const auto insert_index = position - this->begin();
+    const auto insert_back_index = this->end() - position;
+    const auto old_size = static_cast<difference_type>(size());
+    while (first != last) {
+      emplace_front(*first++);
+    }
+    const auto new_position = this->begin() + insert_index;
+    const auto new_position_end = this->end() - insert_back_index;
+    const auto old_begin = this->end() - old_size;
+    std::reverse(this->begin(), old_begin);
+    std::rotate(this->begin(), old_begin, new_position_end);
+    return new_position;
+  }
+}
+
+template <typename T, size_t Capacity>
+auto StaticDeque<T, Capacity>::insert(const const_iterator position,
+                                      const std::initializer_list<T> init)
+    -> iterator {
+  return insert(position, init.begin(), init.end());
+}
+
+template <typename T, size_t Capacity>
+void StaticDeque<T, Capacity>::pop_front() {
+  ASSERT(not this->empty(), "Container is empty.");
+  if (start_position_ == Capacity - 1) {
+    start_position_ = 0;
+  } else {
+    ++start_position_;
+  }
+  --size_;
+  entry_address(Capacity - 1)->~T();
+}
+
+template <typename T, size_t Capacity>
+void StaticDeque<T, Capacity>::pop_back() {
+  ASSERT(not this->empty(), "Container is empty.");
+  --size_;
+  entry_address(size())->~T();
+}
+
+template <typename T, size_t Capacity>
+auto StaticDeque<T, Capacity>::erase(const const_iterator position)
+    -> iterator {
+  return erase(position, position + 1);
+}
+
+template <typename T, size_t Capacity>
+auto StaticDeque<T, Capacity>::erase(const const_iterator first,
+                                     const const_iterator last) -> iterator {
+  const auto result_offset = first - this->begin();
+  auto num_to_remove = last - first;
+  if (num_to_remove == 0) {
+    return last - this->begin() + this->begin();
+  }
+  if (should_operate_on_back(first, last)) {
+    std::move(last, this->cend(), first - this->begin() + this->begin());
+    while (num_to_remove-- != 0) {
+      pop_back();
+    }
+  } else {
+    std::move_backward(this->cbegin(), first,
+                       last - this->begin() + this->begin());
+    while (num_to_remove-- != 0) {
+      pop_front();
+    }
+  }
+  return this->begin() + result_offset;
+}
+
+template <typename T, size_t Capacity>
+void StaticDeque<T, Capacity>::swap(StaticDeque& other) {
+  using std::swap;
+  swap(*this, other);
+}
+
+template <typename T, size_t Capacity>
+void StaticDeque<T, Capacity>::clear() {
+  while (not this->empty()) {
+    pop_back();
+  }
+}
+
+template <typename T, size_t Capacity>
+void StaticDeque<T, Capacity>::pup(PUP::er& p) {
+  size_type sz = size();
+  p | sz;
+  resize(sz);
+  for (auto& entry : *this) {
+    p | entry;
```

Comment:
If we end up with this class holding a fundamental types (I currently don't see that we would get that a lot), then we would want this to use `PUP_array` that serializes a byte stream directly (might be more efficient). I don't think that's important now, but just want to mention it :)

Reply:
I agree with you! Thanks for adding the nice comment!